File: golang/205.isomorphic-strings.go
/*
 * @lc app=leetcode id=205 lang=golang
 *
 * [205] Isomorphic Strings
 */

// @lc code=start
func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	mmap := make(map[rune]rune)
	for i, c := range s {
		if val, exists := mmap[c]; exists {
			if val != rune(t[i]) {
				return false
			}
		} else {
			for _, v := range mmap {
				if v == rune(t[i]) {
					return false
				}
			}
			mmap[c] = rune(t[i])
		}
	}

	return true
}

// @lc code=end

File: python/205.isomorphic-strings.py
#
# @lc app=leetcode id=205 lang=python3
#
# [205] Isomorphic Strings
#


# @lc code=start
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        mmap = {}
        for i, c in enumerate(s):
            if c in mmap:
                if mmap[c] != t[i]:
                    return False

            elif t[i] in mmap.values():
                return False

            mmap[c] = t[i]

        return True


# @lc code=end
